### PR TITLE
[k8s] Parallelize multi-node setup

### PR DESCRIPTION
**Pull request by romilbhardwaj**
Supersedes #4261 and #4270. 

We identified that `apt update` is the slowest operation and triggering it after setup takes a lot of time, so we now inline it in container init args to let kubernetes run it in parallel.

## Testing on nemo image
`sky launch -y -c test --num-nodes 100 --cloud kubernetes --image-id  nvcr.io/nvidia/nemo:24.05.01`

Master branch: 19:56.21 total

This branch: 15:26.56 total

## Testing on default image:
`sky launch -y -c test --num-nodes 100 --cloud kubernetes`

This branch: 4:13.41 total

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] Manual testing with 100 nodes on nemo image, our base image and miniconda image.
- [ ] Need to run relevant smoke tests.

### CODE REVIEW

**Review by Michaelvll — COMMENTED** [review]
Thanks for making the change @romilbhardwaj! Looks mostly good to me.

File: sky/provision/kubernetes/instance.py [hunk 1]
```diff
@@ -456,24 +442,65 @@ def _setup_ssh_in_pods(namespace: str, context: Optional[str],
...
+    # make sure these env vars are available in every task and ssh session.
+    # This is needed for GPU support and service discovery.
+    # See https://github.com/skypilot-org/skypilot/issues/2287 for more details.
+    # To do so, we capture env vars from the pod's runtime and write them to
+    # /etc/profile.d/, making them available for all users in future
+    # shell sessions.
+    set_k8s_env_var_cmd = docker_utils.SETUP_ENV_VARS_CMD
+
+    if not provider_config.get('disable_ssh', False):
+        ssh_install_cmd = install_ssh_k8s_cmd
```
[the remark this answers is not in the document]
Comment (romilbhardwaj) [reply sky/provision/kubernetes/instance.py]: Ah I forgot to push my changes last night. The `disable_ssh` flag has been removed now.

File: sky/provision/kubernetes/instance.py [hunk 2]
```diff
@@ -333,109 +333,94 @@ def _run_function_with_retries(func: Callable,
...
+        '  cat /tmp/apt-update.log; '
+        '  current_time=$(date +%s); '
+        '  elapsed=$((current_time - start_time)); '
+        '  if [ $elapsed -ge $timeout_secs ]; then '
+        '    echo "Timed out waiting for apt update"; '
+        '    exit 1; '
+        '  fi; '
+        '  sleep 5; '
+        'done; '
+        'echo "apt update complete."; ')
```
Comment (Michaelvll) [inline sky/provision/kubernetes/instance.py:382]: nit: how about move this down to keep the order of definition and the order we use them the same for slightly better readability.

File: sky/templates/kubernetes-ray.yml.j2 [hunk 3]
```diff
@@ -417,7 +426,18 @@ setup_commands:
   # Line 'sudo grep ..': set the number of threads per process to unlimited to avoid ray job submit stucking issue when the number of running ray jobs increase.
   # Line 'mkdir -p ..': disable host key check
   # Line 'python3 -c ..': patch the buggy ray files and enable `-o allow_other` option for `goofys`
-  - sudo DEBIAN_FRONTEND=noninteractive apt install lsof gcc patch pciutils rsync fuse curl -y;
+  - |
+    PACKAGES="lsof gcc patch pciutils rsync fuse curl";
```
Comment (Michaelvll) [inline sky/templates/kubernetes-ray.yml.j2:430]: nit: is `lsof` still needed if we are going to remove it?

**Review by Michaelvll — APPROVED** [review]
Thanks for the update @romilbhardwaj! LGTM.

File: sky/templates/kubernetes-ray.yml.j2 (hunk 3 above)
Comment (romilbhardwaj) [reply sky/templates/kubernetes-ray.yml.j2]: Good point, removed now.

**Pull request merged**